chips/esp32/src/timg.rs

```diff
@@ -0,0 +1,232 @@
+        T1 OFFSET(1) NUMBITS(1) [],
+        WDT OFFSET(2) NUMBITS(1) [],
+    ],
+];
+
+pub struct TimG<'a> {
+    registers: StaticRef<TimgRegisters>,
+    alarm_client: OptionalCell<&'a dyn time::AlarmClient>,
+}
+
+impl TimG<'_> {
+    pub const fn new(base: StaticRef<TimgRegisters>) -> Self {
+        TimG {
+            registers: base,
+            alarm_client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let _ = self.stop();
+        self.alarm_client.map(|client| {
+            client.alarm();
+        });
+    }
+
+    pub fn disable_wdt(&self) {
+        self.registers
+            .wdtconfig0
+            .modify(WDTCONFIG0::EN::CLEAR + WDTCONFIG0::FLASHBOOT_MOD_EN::CLEAR);
+
+        if self.registers.wdtconfig0.is_set(WDTCONFIG0::EN)
+            || self
+                .registers
+                .wdtconfig0
+                .is_set(WDTCONFIG0::FLASHBOOT_MOD_EN)
+        {
+            panic!("Can't disable TIMG WDT");
+        }
+    }
+}
+
+impl time::Time for TimG<'_> {
+    type Frequency = Freq80MHz;
+    type Ticks = Ticks64;
+
+    fn now(&self) -> Self::Ticks {
+        self.registers.t0update.set(0xABC);
+        Self::Ticks::from(
+            self.registers.t0lo.get() as u64 + ((self.registers.t0hi.get() as u64) << 32),
+        )
+    }
+}
+
+impl<'a> Counter<'a> for TimG<'a> {
+    fn set_overflow_client(&'a self, _client: &'a dyn time::OverflowClient) {
+        // We have no way to know when this happens
+    }
+
+    fn start(&self) -> Result<(), ErrorCode> {
+        self.registers.t0config.write(CONFIG::EN::SET);
+
+        Ok(())
+    }
+
+    fn stop(&self) -> Result<(), ErrorCode> {
+        self.registers.t0config.write(CONFIG::EN::CLEAR);
+
+        Ok(())
+    }
+
+    fn reset(&self) -> Result<(), ErrorCode> {
+        Err(ErrorCode::FAIL)
+    }
+
+    fn is_running(&self) -> bool {
+        self.registers.t0config.is_set(CONFIG::EN)
+    }
+}
+
+impl<'a> Alarm<'a> for TimG<'a> {
+    fn set_alarm_client(&self, client: &'a dyn time::AlarmClient) {
+        self.alarm_client.set(client);
+    }
+
+    fn set_alarm(&self, _reference: Self::Ticks, _dt: Self::Ticks) {
+        panic!("Unimplemented");
+    }
+
+    fn get_alarm(&self) -> Self::Ticks {
+        panic!("Unimplemented");
+    }
+
+    fn disarm(&self) -> Result<(), ErrorCode> {
+        panic!("Unimplemented");
+    }
+
+    fn is_armed(&self) -> bool {
+        panic!("Unimplemented");
+    }
+
+    fn minimum_dt(&self) -> Self::Ticks {
+        Self::Ticks::from(1 as u64)
+    }
+}
+
+impl kernel::SchedulerTimer for TimG<'_> {
+    fn start(&self, _us: u32) {
+        panic!("Unimplemented");
+    }
+
+    fn reset(&self) {
+        panic!("Unimplemented");
+    }
+
+    fn arm(&self) {
+        panic!("Unimplemented");
+    }
+
+    fn disarm(&self) {
+        panic!("Unimplemented");
+    }
+
+    fn get_remaining_us(&self) -> Option<u32> {
+        panic!("Unimplemented");
+    }
+}
```

Comment:
Do we need this?

Reply:
I would recommend using the virtual scheduler timer instead of this approach